contrib/ruby/ext/trilogy-ruby/cext.c

```diff
@@ -1095,6 +1147,45 @@ static VALUE rb_trilogy_write_timeout_set(VALUE self, VALUE write_timeout)
     return write_timeout;
 }
 
+const char *const trilogy_type_names[] = {
+#define XX(name, code) [name] = (char *)#name + strlen("TRILOGY_TYPE_"),
+    TRILOGY_TYPES(XX)
+#undef XX
+};
+
+const char *const trilogy_charset_names[] = {
+#define XX(name, code) [name] = (char *)#name + strlen("TRILOGY_CHARSET_"),
+    TRILOGY_CHARSETS(XX)
+#undef XX
+};
+
+static char *downcase(const char *str)
```

Comment:
This is to convert the charset/type name into a lowercase symbol. I'm sure there is a better way to do this.